services/sensiolabs/sensiolabs.service.js

```diff
@@ -1,110 +1,258 @@
-module.exports = class Sensiolabs extends LegacyService {
+const schema = Joi.object({
+  project: Joi.object({
+    'last-analysis': Joi.object({
+      status: Joi.string()
+        .regex(/ordered|running|measured|analyzed|finished/)
+        .required(),
+      grade: Joi.string()
+        .regex(/platinum|gold|silver|bronze|none/)
+        .allow('', null),
+      violations: Joi.object({
+        violation: Joi.array()
+          .optional()
+          .items(
+            Joi.object({
+              severity: Joi.string()
+                .regex(/info|minor|major|critical/)
+                .required(),
+            })
+          ),
+      })
+        .optional()
+        .allow(''),
+    }),
+  }).required(),
+}).required()
+
+module.exports = class Sensiolabs extends BaseXmlService {
+  static render({
+    metric,
+    status,
+    grade,
+    numViolations,
+    numCriticalViolations,
+    numMajorViolations,
+    numMinorViolations,
+    numInfoViolations,
+  }) {
+    if (status !== 'finished') {
+      return {
+        message: 'pending',
+        color: 'grey',
+      }
+    }
+
+    if (metric === 'grade') {
+      return this.renderGradeBadge({ grade })
+    } else {
+      return this.renderViolationsBadge({
+        numViolations,
+        numCriticalViolations,
+        numMajorViolations,
+        numMinorViolations,
+        numInfoViolations,
+      })
+    }
+  }
+
+  static renderGradeBadge({ grade }) {
+    let color,
+      message = grade
+    if (grade === 'platinum') {
+      color = 'brightgreen'
+    } else if (grade === 'gold') {
+      color = 'yellow'
+    } else if (grade === 'silver') {
+      color = 'lightgrey'
+    } else if (grade === 'bronze') {
+      color = 'orange'
+    } else {
+      message = 'no medal'
+      color = 'red'
+    }
+
+    return {
+      message,
+      color,
+    }
+  }
+
+  static renderViolationsBadge({
+    numViolations,
+    numCriticalViolations,
+    numMajorViolations,
+    numMinorViolations,
+    numInfoViolations,
+  }) {
+    if (numViolations === 0) {
+      return {
+        message: '0 violations',
+        color: 'brightgreen',
+      }
+    }
+
+    let color = 'yellowgreen'
+    const violationSummary = []
+
+    if (numInfoViolations > 0) {
+      violationSummary.push(`${numInfoViolations} info`)
+    }
+    if (numMinorViolations > 0) {
+      violationSummary.unshift(`${numMinorViolations} minor`)
+      color = 'yellow'
+    }
+    if (numMajorViolations > 0) {
+      violationSummary.unshift(`${numMajorViolations} major`)
+      color = 'orange'
+    }
+    if (numCriticalViolations > 0) {
+      violationSummary.unshift(`${numCriticalViolations} major`)
+      color = 'red'
+    }
+
+    return {
+      message: violationSummary.join(', '),
+      color,
+    }
+  }
+
   static get category() {
-    return 'build'
+    return 'quality'
   }
 
   static get route() {
     return {
-      base: 'sensiolabs',
+      // base: '(symfony|sensiolabs)/i',
```

Comment:
I could not get any sort of pattern/regex to work in `base`, so if there's a way to do so please let me know! In the absence of a dynamic `base` pattern, I had two ideas for approaches that would allow us to update the badge accordingly to reflect the product rebrand to SymfonyInsight without breaking existing users of the `sensiolabs` badge:

1. Stick with a single service class (which I still need to rename, etc.) and ensure the `pattern` would support the old path `sensiolabs` as well as the new path `symfony`
2. Have two service classes, one for `symfony` that would contain the core implementation, and then another for `sensiolabs` that could extend the `symfony` class to reuse that logic

Reply:
I'm not sure about the rest of the pattern, though if you try setting `base` to `''` you can move the dynamic prefix into the format / pattern. We do that with azure devops.

Btw, in the medium term I'd like to solve the aliasing problem using a redirect (#2340).

Reply:
I ended up doing just that, adding it to the pattern:

```js
  pattern: '(symfony|sensiolabs)/i/:metric(grade|violations)?/:projectUuid'
```

I also made the metric optional (and default to `grade` when not provided) as the old path was simply `sensiolabs/i/:projectUuid`

What's the pattern for redirects with the new service model?

Reply:
(Totally misread that the first time. Here's take 2!)

Yea, you can't use a pattern in `base`. It doesn't really make sense for it to be dynamic because of the way we use it in testers and examples.

At some point I'd like to solve the aliasing problem using a redirect (#2340). If we can get #2519 merged we could unblock that, though I'm not sure it's a higher priority than finishing up the rewrite. So I'd probably suggest option 1 for now, which is also what we did in azure devops, and then revisiting when we get some redirect capability in core.

Reply:
I wonder if we could do something like this (or its regex equivalent) to avoid providing new API surface area:

```js
pattern: '(sensiolabs/i|symfony/i/:metric(grade|violations))/:projectUuid'
```

If we never start supporting `sensiolabs/i/violations` it's less that needs to be redirected later.

Reply:
gotcha, that's a good idea. will do